src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -369,42 +459,27 @@ public bool TryGetValue(TKey key, [MaybeNullWhen(false)] out TValue value)
                 ThrowHelper.ThrowKeyNullException();
             }
 
-            // We must capture the volatile _tables field into a local variable: it is set to a new table on each table resize.
-            // The Volatile.Read on the array element then ensures that we have a copy of the reference to tables._buckets[bucketNo]:
-            // this protects us from reading fields ('_hashcode', '_key', '_value' and '_next') of different instances.
             Tables tables = _tables;
 
-            IEqualityComparer<TKey>? comparer = _comparer;
-            if (comparer is null)
+            IEqualityComparer<TKey>? comparer = tables._comparer;
+            if (typeof(TKey).IsValueType && // comparer can only be null for value types; enable JIT to eliminate entire if block for ref types
+                comparer is null)
             {
                 int hashcode = key.GetHashCode();
-                if (typeof(TKey).IsValueType)
+                for (Node? n = GetBucket(tables, hashcode); n is not null; n = n._next)
                 {
-                    for (Node? n = Volatile.Read(ref tables.GetBucket(hashcode)); n != null; n = n._next)
+                    if (hashcode == n._hashcode && EqualityComparer<TKey>.Default.Equals(n._key, key))
                     {
-                        if (hashcode == n._hashcode && EqualityComparer<TKey>.Default.Equals(n._key, key))
-                        {
-                            value = n._value;
-                            return true;
-                        }
-                    }
-                }
-                else
-                {
-                    for (Node? n = Volatile.Read(ref tables.GetBucket(hashcode)); n != null; n = n._next)
-                    {
-                        if (hashcode == n._hashcode && _defaultComparer.Equals(n._key, key))
-                        {
-                            value = n._value;
-                            return true;
-                        }
+                        value = n._value;
+                        return true;
                     }
                 }
             }
             else
             {
-                int hashcode = comparer.GetHashCode(key);
-                for (Node? n = Volatile.Read(ref tables.GetBucket(hashcode)); n != null; n = n._next)
+                Debug.Assert(comparer is not null);
+                int hashcode = _comparerIsDefaultForClasses ? key.GetHashCode() : comparer.GetHashCode(key);
```

Comment:
Nit: This line looks same as code in `GetHashCode(IEqualityComparer<TKey>? comparer, TKey key)`.

Reply:
Yup, I'll dedup that, thanks.